### PR TITLE
[PROF-1988] default stackdepth

### PR DESCRIPTION
**Pull request by pmbauer**
Change Overview
------------------

* sets JFR profiles to use a 256 stack depth by default
* respects user-supplied over-ride for max stack depth
* limits clients to a 1024 max stack depth
* jboss (wildfly) integration tests

Observations
------------------
Based on [benchmark results](https://docs.google.com/spreadsheets/d/1E0ZuxqwrLXIK0w0n6n6ZZ9ASZ029lToB04SWK7OAW_I/edit#gid=831612831) there is little overhead increasing the `stackdepth` and not truncating typical java web application stacks, which can be regularly greater than 128.

Due to quirks in JFR initialization ([one-shot per JVM, sampling thread started as a side-effect of setting the sampling interval](https://github.com/openjdk/jdk/blob/e0cf023263fcd8ad5c8235828e025c5842807bfe/src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp#L596)) and issues initializing JMX in some environments, there is a small window in which the agent can set `JFR.configure stackdepth`
1. before profiling starts
2. after JMX is initialized

As a result, it's possible the agent may fail to set `stackdepth` appropriately in some environments.  That occurrence is logged for diagnostic purposes.  The product will still function as before but some clients may get a warning in the analysis tab (as they do today) that their stacks were truncated with mediation steps.

[PROF-1988]

[PROF-1988]: https://datadoghq.atlassian.net/browse/PROF-1988

### CODE REVIEW

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java [hunk 1]
```diff
@@ -126,9 +131,53 @@ public final void start() {
     }
   }
 
+  private static void setMaxStackDepth() {
+    final String JFR_STACK_DEPTH_ARG = "-XX:FlightRecorderOptions=stackdepth=";
```
Comment (jpbempel) [inline dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java:135]: you may have stackdepth as second option like:
`-XX:FlightRecorderOptions=dumponexit=true,stackdepth=128`

Comment (pmbauer) [reply dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java:135]: good point, will fix

File: dd-smoke-tests/wildfly/src/test/groovy/datadog/smoketest/JFRStackDepthTest.groovy [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    processBuilder.environment().put("JAVA_OPTS", javaOpts.join(" "))
+    processBuilder.environment().put("JAVA_HOME", System.getProperty("java.home"))
+    processBuilder.environment().put("DD_API_KEY", "01234567890abcdef123456789ABCDEF")
+    processBuilder.redirectErrorStream(true)
+    processBuilder.redirectOutput(ProcessBuilder.Redirect.to(new File(logFilePath)))
+    testedProcess = processBuilder.start()
+
+    PortUtils.waitForPortToOpen(httpPort, 240, TimeUnit.SECONDS, testedProcess)
+
+    // This test can't run on Java 1.8, requires Process::children and Process::pid only present since 1.9
```
Comment (pmbauer) [inline dd-smoke-tests/wildfly/src/test/groovy/datadog/smoketest/JFRStackDepthTest.groovy:93]: is there a better way to do this with gradle?

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java [hunk 3]
```diff
@@ -126,9 +132,73 @@ public final void start() {
...
+
+    // don't set stackdepth if the client has explicitly set it
+    final Optional<String> userSpecifiedStackDepth =
+        readJFRStackDepth(SystemAccess.getVMArguments());
+    if (userSpecifiedStackDepth.isPresent()) {
+      maxFrames = stackDepthFromClient(userSpecifiedStackDepth.get());
+    }
+
+    log.info("setting JFR.configure stackdepth=" + maxFrames);
+    SystemAccess.executeDiagnosticCommand(
```
Comment (jpbempel) [inline dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java:166]: Can we detect in one way or another that JMX is not yet initialized when we performed the call?
to help diganostic when it does not work. 
return of the method is null = meaning this is the NoOp?

**Review by jpbempel — APPROVED** [review]

File: dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java (hunk 3 above)
Comment (pmbauer) [reply dd-java-agent/agent-profiling/profiling-controller/src/main/java/com/datadog/profiling/controller/ProfilingSystem.java]: fixed, reporting results https://github.com/DataDog/dd-trace-java/pull/1928/commits/df41df5043ba296875390c8a35368740888d20e9

**Review by richardstartin — APPROVED** [review]

**Pull request merged**